### ISSUE TITLE
Log training-time metrics to MLflow during fine-tuning (report_to is disabled) — needed to demonstrate MLflow experiment tracking

### ISSUE THREAD

**Issue by hiouchiy** [hiouchiy/air-sample-bert#10]
## Log training-time metrics to MLflow (currently disabled) — this is the core Databricks value to show

**Current state:** both fine-tuning notebooks set `TrainingArguments(..., report_to=[])`:
- `01_finetune_singlegpu.py` L230
- `02_finetune_multigpu.py` L208

This **disables HuggingFace Trainer's built-in MLflow callback**, so the per-step training **loss** / learning-rate and the per-epoch **eval_loss / accuracy / f1** are only *printed* — they are **never logged to the MLflow experiment**. Only the final metrics + params + model are logged afterward in `log_and_register`. The MLflow run therefore has **no training curves** — which is exactly the artifact that showcases MLflow experiment tracking, a headline Databricks value.

**Why it's likely disabled:** `log_and_register` opens its own `mlflow.start_run()`, and the Trainer's callback would otherwise create a *separate* run (duplicate / nesting). Understandable, but the cost is losing the training curves.

**Suggested fix — one run with curves + params + model:**
```python
# TrainingArguments: send Trainer metrics to MLflow
report_to=["mlflow"],
```
```python
# main(): wrap training AND registration in the SAME run
with mlflow.start_run(run_name="modernbert-agnews-singlegpu"):
    trainer, metrics = train(CFG, tokenized, tokenizer)   # MLflowCallback logs loss/eval
                                                          # curves to THIS active run
    log_and_register(CFG, trainer, tokenizer, metrics)     # params + final metrics + model,
                                                          # same run  (drop its own start_run)
```
- HF's `MLflowCallback` logs into the **already-active** run, so opening the run first avoids the duplicate/nested run that `report_to=[]` was avoiding.
- DDP (`02`): the callback logs from **rank 0 only**, so it's safe.
- Result: one run showing the **training/eval curves → final metrics → registered UC model** — the end-to-end "track it in MLflow, register to UC, serve" story.

*(Companion, milder: the XGBoost sample logs only final metrics too; its per-round `evals_result` curve could be logged via `mlflow.log_metrics(..., step=i)`. Separate repo.)*

**Comment by hiouchiy** [comment]
Fixed in bb64bbd. Fine-tuning now uses report_to=["mlflow"] and opens ONE run before train(), so HF's callback streams the loss/eval curves into the same run that gets the final metrics + registered model (02 does this on rank 0). Custom params use distinct keys and are logged defensively to avoid colliding with the callback's model-config params (e.g. max_length). **Validated E2E on A10**: run has the training-loss/eval curves + accuracy/f1 + registered v10 @champion.

**Comment by hiouchiy** [comment]
## Clarification: the §5-train / §6-log split is *why* the run can't hold the training curve — open the run around training

A reviewer asked why `01`/`02` split "5. Train" from "6. Log to MLflow & register" into separate cells/functions, and whether logging the run + model artifact right in the Train step would be more natural. Answer: the split is **only stylistic** (separating "train" from "register to UC" as teaching steps; it also lets you re-register without retraining). There's **no technical requirement** for it — and it's actually the root of this issue:

- Today `train()` runs **outside any MLflow run**, and `log_and_register()` opens `mlflow.start_run()` **afterward** (§6). So the HF Trainer's per-step metrics (once `report_to="mlflow"` is enabled) have **no active run to land in** — they can't join the run that gets the model.
- The natural, correct structure is to **open one run that spans training and registration** so the single run holds **training curve + params + final metrics + model artifact**:
  ```python
  with mlflow.start_run(run_name=...):
      trainer, metrics = train(...)       # report_to="mlflow" logs the curve into THIS run
      log_and_register(...)               # same run: params + final metrics + log_model + @champion
  ```
- This can still be presented as two cells if desired, but the **run must be opened before the Train cell** (a run spanning cells) — or simply keep training + logging under one `with mlflow.start_run()`. Either way, don't open the run only in §6.

So the fix for this issue also resolves the "why split §5/§6?" question: **wrap training in the run** rather than logging after the fact. Applies to `01_finetune_singlegpu` and `02_finetune_multigpu` (rank-0).
